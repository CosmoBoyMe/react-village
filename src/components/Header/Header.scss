@import '../../styles/mixins';

.header {
  background-color: #fff;
  box-shadow: 0 10px 20px var(--white);

  &__container {
    max-width: 1200px;
    margin: 0 auto;
    padding: 0 20px;
  }

  &__inner {
    display: flex;
    align-items: center;
    padding: 15px 0;
  }

  &__nav {
    display: flex;
    align-items: center;
    margin-left: auto;

    &_active {
      position: absolute;
      top: 0;
      right: 0;
      z-index: 999;
      padding: 20px;
      background-color: #fff;
      border-bottom-left-radius: 6px;

      @include medium {
        position: fixed;
        align-items: flex-start;
        justify-content: center;
        width: 100%;
        height: 100%;
        overflow: scroll;
        border-bottom-left-radius: 0;
      }
    }

    &-burger {
      --burger-width: 40px;
      --burger-height: 20px;
      --burger-line-height: 2px;

      position: absolute;
      top: 20px;
      right: 20px;
      z-index: 1000;
      display: none;
      width: var(--burger-width);
      height: var(--burger-height);
      padding: 0;
      color: var(--dark-shade-50);
      background-color: transparent;
      border: none;
      cursor: pointer;

      &::before,
      &::after {
        position: absolute;
        left: 0;
        width: 100%;
        height: var(--burger-line-height);
        background-color: currentcolor;
        transition: transform 0.3s ease-in-out, top 0.3s ease-in-out;
        content: '';
      }

      &::before {
        top: 0;
      }

      &::after {
        top: calc(100% - var(--burger-line-height));
      }

      &_active::before {
        top: 50%;
        transform: rotate(45deg);
        transition: transform 0.3s ease-in-out, top 0.3s ease-in-out;
      }

      &_active::after {
        top: 50%;
        transform: rotate(-45deg);
        transition: transform 0.3s ease-in-out, top 0.3s ease-in-out;
      }

      &-line {
        position: absolute;
        top: 50%;
        left: 0;
        width: 100%;
        height: var(--burger-line-height);
        background-color: currentcolor;
        transform: translateY(-50%);
        transition: transform 0.3s ease-in-out;
      }

      &_active &-line {
        transform: scale(0);
        transition: transform 0.3s ease-in-out;
      }

      &:focus::after {
        opacity: 1;
      }

      @include large {
        display: block;
      }
    }

    &-content {
      display: flex;
      align-items: center;

      @include large {
        display: none;
        flex-direction: column;
        align-items: stretch;
      }
    }

    &_active &-content {
      display: flex;

      @include medium {
        align-items: center;
      }
    }

    &-list {
      display: flex;
      column-gap: 20px;
      list-style: none;

      @include large {
        flex-direction: column;
        margin-bottom: 20px;
        row-gap: 20px;
      }
    }

    &-link {
      display: inline-block;
      color: var(--dark-shade-50);
      line-height: 17px;
      text-decoration: none;

      &::first-letter {
        text-transform: uppercase;
      }

      &_active {
        color: var(--dark-shade-75);
        font-weight: 700;
      }

      &_sign_arrow {
        position: relative;
        padding-right: 25px;

        &::after {
          top: -3px;
          right: 0;

          @include material-icon('expand_more');
        }
      }

      &:hover {
        opacity: 0.5;
      }
    }

    &-profile {
      position: relative;
      margin-left: 20px;
      padding-left: 20px;
      color: var(--dark-shade-50);
      text-transform: capitalize;
      cursor: pointer;

      &:hover {
        color: var(--dark-shade);
      }

      &::before {
        position: absolute;
        top: 0;
        left: 0;
        width: 1px;
        height: 100%;
        background: rgb(31 32 65 / 10%);
        content: '';
      }
    }

    &-auth-user {
      display: flex;
      margin-bottom: 1px;
      margin-left: 20px;
      column-gap: 20px;

      @include large {
        flex-direction: column;
        max-width: 300px;
        margin-left: 0;
        row-gap: 20px;
      }
    }
  }

  @include large {
    position: relative;
  }
}
