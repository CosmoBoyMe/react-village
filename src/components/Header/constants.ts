import SCREENS from '../../routes/endpoints';

const navigationItems = [
  { text: 'о нас', to: SCREENS.MAIN },
  { text: 'услуги', isNested: true, to: '/mock-address/change-me' },
  { text: 'вакансии', to: '/mock-address/change-me' },
  { text: 'новости', to: '/mock-address/change-me' },
  { text: 'соглашения', isNested: true, to: '/mock-address/change-me' },
];

export { navigationItems };
