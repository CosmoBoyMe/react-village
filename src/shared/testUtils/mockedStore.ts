import { initialState as authInitialState } from '../../store/slices/auth/slice';
import { initialState as bookingInitialState } from '../../store/slices/booking/slice';
import { initialState as filtersInitialState } from '../../store/slices/filters/slice';
import { initialState as profileInitialState } from '../../store/slices/profile/slice';
import { initialState as roomInitialState } from '../../store/slices/room/slice';
import { initialState as roomsInitialState } from '../../store/slices/rooms/slice';

const mockedStore = {
  auth: authInitialState,
  filters: filtersInitialState,
  rooms: roomsInitialState,
  room: roomInitialState,
  booking: bookingInitialState,
  profile: profileInitialState,
};

export { mockedStore };
