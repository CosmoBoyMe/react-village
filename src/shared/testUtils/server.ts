import { rest } from 'msw';
import { setupServer } from 'msw/node';

const server = setupServer(
  rest.post(
    'https://react-village-d5bce-default-rtdb.firebaseio.com/rooms/1/bookedDates.json',
    (req, res, ctx) => {
      return res(ctx.status(200), ctx.json({ bookingID: '123' }));
    }
  ),

  rest.post(
    'https://react-village-d5bce-default-rtdb.firebaseio.com/rooms/1/feedback.json',
    (req, res, ctx) => {
      return res(ctx.status(200), ctx.json({}));
    }
  ),

  rest.post(
    'https://react-village-d5bce-default-rtdb.firebaseio.com/users/testUser/booking.json',
    (req, res, ctx) => {
      return res(ctx.status(200));
    }
  ),
  rest.get(
    'https://react-village-d5bce-default-rtdb.firebaseio.com/rooms.json',
    (req, res, ctx) => {
      return res(
        ctx.status(200),
        ctx.json({
          1: {},
        })
      );
    }
  )
);

export { server };
