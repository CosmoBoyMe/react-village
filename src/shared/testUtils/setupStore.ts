import type { PreloadedState } from '@reduxjs/toolkit';
import { configureStore } from '@reduxjs/toolkit';

import { rootReducer } from '../../store/index';

type RootState = ReturnType<typeof rootReducer>;

const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        serializableCheck: false,
      }),
  });
};

type AppStore = ReturnType<typeof setupStore>;
type AppDispatch = AppStore['dispatch'];

export { setupStore };
export type { AppDispatch, AppStore, RootState };
