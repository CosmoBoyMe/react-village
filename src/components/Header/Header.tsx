import { FC, memo, useCallback, useEffect, useRef, useState } from 'react';
import { useSelector } from 'react-redux';
import { Link, NavLink, useNavigate } from 'react-router-dom';
import classNames from 'classnames';

import { SCREENS } from '../../routes/endpoints';
import { WindowSizes } from '../../shared/constants/WindowSizes';
import {
  isAuthSelect,
  userNameSelect,
  userSurnameSelect,
} from '../../store/slices/auth/selectors';
import { ButtonLink } from '../ButtonLink/ButtonLink';
import { Logo } from '../Logo/Logo';

import { navigationItems } from './constants';
import './Header.scss';

const Header: FC = memo(() => {
  const navigationRef = useRef<HTMLDivElement>(null);
  const navigate = useNavigate();
  const isAuth = useSelector(isAuthSelect);
  const userName = useSelector(userNameSelect);
  const userSurname = useSelector(userSurnameSelect);

  const [isBurgerMenuActive, setIsBurgerMenuActive] = useState(false);
  const handleNavBurgerClick = () => {
    setIsBurgerMenuActive(!isBurgerMenuActive);

    if (window.screen.width <= WindowSizes.Medium) {
      document.body.style.overflow = isBurgerMenuActive ? '' : 'hidden';
    }
  };

  const handleLinkClick = useCallback(() => {
    setIsBurgerMenuActive(false);
  }, []);

  const handleBodyClick = ({ currentTarget, target }: MouseEvent) => {
    if (
      currentTarget instanceof HTMLElement &&
      target instanceof HTMLElement &&
      navigationRef.current !== null
    ) {
      if (
        currentTarget.offsetWidth > WindowSizes.Medium &&
        currentTarget.offsetWidth <= WindowSizes.Large
      ) {
        if (!navigationRef.current.contains(target)) {
          setIsBurgerMenuActive(false);
        }
      }
    }
  };

  const handleUserNavProfileClick = () => {
    navigate(SCREENS.PROFILE);
  };

  useEffect(() => {
    const handleWindowResize = () => {
      if (
        window.screen.width > WindowSizes.Medium &&
        window.screen.width <= WindowSizes.Large
      ) {
        if (isBurgerMenuActive) {
          document.body.style.overflow = '';
        }
        document.body.addEventListener('click', handleBodyClick);
      } else if (window.screen.width <= WindowSizes.Medium) {
        if (isBurgerMenuActive) {
          document.body.style.overflow = 'hidden';
        }
        document.body.removeEventListener('click', handleBodyClick);
      } else if (window.screen.width > WindowSizes.Large) {
        if (isBurgerMenuActive) {
          document.body.style.overflow = '';
        }
        setIsBurgerMenuActive(false);
      }
    };

    window.addEventListener('resize', handleWindowResize);
    return () => {
      window.removeEventListener('resize', handleWindowResize);
    };
  }, [isBurgerMenuActive]);

  useEffect(() => {
    if (
      window.screen.width > WindowSizes.Medium &&
      window.screen.width <= WindowSizes.Large
    ) {
      document.body.addEventListener('click', handleBodyClick);
    }
    return () => {
      document.body.removeEventListener('click', handleBodyClick);
    };
  }, []);

  return (
    <header className="header">
      <div className="header__container">
        <div className="header__inner">
          <Link
            onClick={handleLinkClick}
            to={SCREENS.LANDING}
            className="header__logo-link"
          >
            <Logo />
          </Link>
          <nav
            ref={navigationRef}
            className={classNames('header__nav', {
              header__nav_active: isBurgerMenuActive,
            })}
          >
            <button
              title="главное меню"
              type="button"
              onClick={handleNavBurgerClick}
              className={classNames('header__nav-burger', {
                'header__nav-burger_active': isBurgerMenuActive,
              })}
            >
              <span className="header__nav-burger-line" />
            </button>
            <div className="header__nav-content">
              <ul className="header__nav-list">
                {navigationItems.map((item) => (
                  <li key={item.text} className="header__nav-item">
                    <NavLink
                      onClick={handleLinkClick}
                      to={item.to}
                      className={({ isActive }) =>
                        classNames('header__nav-link', {
                          'header__nav-link_active':
                            isActive && item.to !== '/mock-address/change-me',
                          'header__nav-link_sign_arrow': item.isNested,
                        })
                      }
                    >
                      {item.text}
                    </NavLink>
                  </li>
                ))}
              </ul>
              {isAuth && userName && userSurname ? (
                <button
                  type="button"
                  onClick={handleUserNavProfileClick}
                  className="header__nav-profile"
                >
                  {`${userName} ${userSurname}`}
                </button>
              ) : (
                <div className="header__nav-auth-user">
                  <ButtonLink
                    onClick={handleLinkClick}
                    text="войти"
                    href={SCREENS.SIGN_IN}
                    withBorder
                    isSmall
                  />
                  <ButtonLink
                    onClick={handleLinkClick}
                    text="зарегистрироваться"
                    href={SCREENS.SIGN_UP}
                    isSmall
                  />
                </div>
              )}
            </div>
          </nav>
        </div>
      </div>
    </header>
  );
});

Header.displayName = 'Header';

export { Header };
