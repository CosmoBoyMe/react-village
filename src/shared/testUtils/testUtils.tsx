/* eslint-disable import/no-extraneous-dependencies */
import React, { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import type { PreloadedState } from '@reduxjs/toolkit';
import { configureStore } from '@reduxjs/toolkit';
import type { RenderOptions } from '@testing-library/react';
import { render } from '@testing-library/react';

import type { RootState } from '../../store';
import { rootReducer } from '../../store';

import { mockedStore } from './mockedStore';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: ReturnType<typeof setupStore>;
}

const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
};

function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = mockedStore,
    store = setupStore(mockedStore),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<unknown>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export { renderWithProviders };
