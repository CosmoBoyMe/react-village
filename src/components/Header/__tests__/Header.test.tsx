/* eslint-disable testing-library/render-result-naming-convention */
import { BrowserRouter } from 'react-router-dom';

import { mockedStore } from '../../../shared/testUtils/mockedStore';
import { renderWithProviders } from '../../../shared/testUtils/testUtils';
import { Header } from '../Header';

jest.mock('axios');

describe('Header component rendering', () => {
  it('Renders Header component', () => {
    const headerComponent = renderWithProviders(
      <BrowserRouter>
        <Header />
      </BrowserRouter>,
      {
        preloadedState: mockedStore,
      }
    );

    expect(headerComponent).toMatchSnapshot();
  });
});
