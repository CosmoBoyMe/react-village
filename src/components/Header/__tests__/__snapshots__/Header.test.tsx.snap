// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component rendering Renders Header component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="header"
      >
        <div
          class="header__container"
        >
          <div
            class="header__inner"
          >
            <a
              class="header__logo-link"
              href="/"
            >
              <div
                class="logo"
              >
                <img
                  alt="company logo"
                  class="logo__img"
                  src="logo.svg"
                />
              </div>
            </a>
            <nav
              class="header__nav"
            >
              <button
                class="header__nav-burger"
                type="button"
              >
                <span
                  class="header__nav-burger-line"
                />
              </button>
              <div
                class="header__nav-content"
              >
                <ul
                  class="header__nav-list"
                >
                  <li
                    class="header__nav-item"
                  >
                    <a
                      aria-current="page"
                      class="header__nav-link header__nav-link_active"
                      href="/"
                    >
                      о нас
                    </a>
                  </li>
                  <li
                    class="header__nav-item"
                  >
                    <a
                      class="header__nav-link header__nav-link_sign_arrow"
                      href="/mock-address/change-me"
                    >
                      услуги
                    </a>
                  </li>
                  <li
                    class="header__nav-item"
                  >
                    <a
                      class="header__nav-link"
                      href="/mock-address/change-me"
                    >
                      вакансии
                    </a>
                  </li>
                  <li
                    class="header__nav-item"
                  >
                    <a
                      class="header__nav-link"
                      href="/mock-address/change-me"
                    >
                      новости
                    </a>
                  </li>
                  <li
                    class="header__nav-item"
                  >
                    <a
                      class="header__nav-link header__nav-link_sign_arrow"
                      href="/mock-address/change-me"
                    >
                      соглашения
                    </a>
                  </li>
                </ul>
                <div
                  class="header__nav-auth-user"
                >
                  <a
                    class="button-link button-link_bordered button-link_size_small"
                    href="/sign-in"
                  >
                    войти
                  </a>
                  <a
                    class="button-link button-link_size_small"
                    href="/sign-up"
                  >
                    зарегистрироваться
                  </a>
                </div>
              </div>
            </nav>
          </div>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="header"
    >
      <div
        class="header__container"
      >
        <div
          class="header__inner"
        >
          <a
            class="header__logo-link"
            href="/"
          >
            <div
              class="logo"
            >
              <img
                alt="company logo"
                class="logo__img"
                src="logo.svg"
              />
            </div>
          </a>
          <nav
            class="header__nav"
          >
            <button
              class="header__nav-burger"
              type="button"
            >
              <span
                class="header__nav-burger-line"
              />
            </button>
            <div
              class="header__nav-content"
            >
              <ul
                class="header__nav-list"
              >
                <li
                  class="header__nav-item"
                >
                  <a
                    aria-current="page"
                    class="header__nav-link header__nav-link_active"
                    href="/"
                  >
                    о нас
                  </a>
                </li>
                <li
                  class="header__nav-item"
                >
                  <a
                    class="header__nav-link header__nav-link_sign_arrow"
                    href="/mock-address/change-me"
                  >
                    услуги
                  </a>
                </li>
                <li
                  class="header__nav-item"
                >
                  <a
                    class="header__nav-link"
                    href="/mock-address/change-me"
                  >
                    вакансии
                  </a>
                </li>
                <li
                  class="header__nav-item"
                >
                  <a
                    class="header__nav-link"
                    href="/mock-address/change-me"
                  >
                    новости
                  </a>
                </li>
                <li
                  class="header__nav-item"
                >
                  <a
                    class="header__nav-link header__nav-link_sign_arrow"
                    href="/mock-address/change-me"
                  >
                    соглашения
                  </a>
                </li>
              </ul>
              <div
                class="header__nav-auth-user"
              >
                <a
                  class="button-link button-link_bordered button-link_size_small"
                  href="/sign-in"
                >
                  войти
                </a>
                <a
                  class="button-link button-link_size_small"
                  href="/sign-up"
                >
                  зарегистрироваться
                </a>
              </div>
            </div>
          </nav>
        </div>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
